_source/_posts/2022-08-31-state-of-java-project-loom.md

```diff
@@ -0,0 +1,263 @@
+    }
+}
+```
+
+We want `updateInventory()` and `updateOrder()` subtasks to be executed concurrently. Each of those can succeed or fail independently. Ideally, the `handleOrder()` method should fail if any subtask fails. However, if a failure occurs in one subtask, things get messy.
+
+- Imagine that `updateInventory()` failed and throws an exception. Then, the `handleOrder()` method will throw an exception when calling `inventory.get()`. So far is fine, but what about `updateOrder()`? Since it runs on its own thread, it can complete successfully. But now we have an issue with a mismatch in inventory and order. Suppose the `updateOrder()` is an expensive operation. In that case, we are just wasting the resources for nothing, and we would have to write some sort of guard logic to revert the updates done to order as our overall operation failed.
+- Imagine that `updateInventory()` is an expensive long-running operation and `updateOrder()` threw an error. The `handleOrder()` task will be blocked on `inventory.get()` even though `updateOrder()` threw an error. Ideally, we would like the `handleOrder()` task to cancel `updateInventory()` when a failure occurs in `updateOrder()` so that we are not wasting time.
+- If the thread executing `handleOrder()` is interrupted, the interruption is not propagated to the subtasks. In this case `updateInventory()` and `updateOrder()` will leak and continue to run in the background.
+
+For these situations, we would have to carefully write workarounds and failsafe, putting all the burden on the developer.
+
+We can achieve the same functionality with structured concurrency using the code below.
+
+```java
+void handleOrder() throws ExecutionException, InterruptedException {
+    try (var scope = new StructuredTaskScope.ShutdownOnFailure()) {
+        Future<Integer> inventory = scope.fork(() -> updateInventory());
+        Future<Integer> order = scope.fork(() -> updateOrder());
+
+        scope.join();           // Join both forks
+        scope.throwIfFailed();  // ... and propagate errors
+
+        // Here, both forks have succeeded, so compose their results
+        System.out.println("Inventory " + inventory.resultNow() + " updated for order " + order.resultNow());
+    }
+}
+```
+
+Unlike the previous sample using `ExecutorService`, we can now use `StructuredTaskScope` to achieve the same result while confining the lifetimes of the subtasks to the lexical scope, in this case, the body of the _try-with-resources_ statement. The code is much more readable, and the intent is also clear. `StructuredTaskScope` also ensures the below behavior automatically
+
+- **Error handling with short-circuiting** — If either the `updateInventory()` or `updateOrder()` fails, the other is cancelled unless its already completed. This is managed by the cancellation policy implemented by `ShutdownOnFailure()`; other policies are possible.
```

Comment:
```suggestion
- **Error handling with short-circuiting** — If either the `updateInventory()` or `updateOrder()` fails, the other is canceled unless its already completed. This is managed by the cancellation policy implemented by `ShutdownOnFailure()`; other policies are possible.
```
both canceled and cancelled are acceptable for the past tense of cancel, the version with one L is more common in American English, while the version with two L's is more common in British English. changing for consistency :-)